File: shankejdk-simple-sources/jdk/src/macosx/native/jobjc/src/core/java/com/apple/jobjc/NativeObjectLifecycleManager.java
/*
 * Copyright (c) 2011, 2013, Oracle and/or its affiliates. All rights reserved.
 */
package com.apple.jobjc;


public abstract class NativeObjectLifecycleManager {
    private static native void retainNativeObject(final long ptr);
    private static native void releaseNativeObject(final long ptr);
    private static native void freeNativeObject(final long ptr);

    abstract void begin(final long ptr);
    abstract void end(final long ptr);
    boolean shouldPreRetain() { return false; }

    public static class CFRetainRelease extends NativeObjectLifecycleManager {
        public static final NativeObjectLifecycleManager INST = new CFRetainRelease();
        @Override void begin(final long ptr) { retainNativeObject(ptr); }
        @Override void end(final long ptr) { releaseNativeObject(ptr); }
        @Override boolean shouldPreRetain() { return true; }
    }

    public static class Free extends NativeObjectLifecycleManager {
        public static final NativeObjectLifecycleManager INST = new Free();
        @Override void begin(final long ptr) { }
        @Override void end(final long ptr) { freeNativeObject(ptr); }
    }

    public static class Nothing extends NativeObjectLifecycleManager {
        public static final NativeObjectLifecycleManager INST = new Nothing();
        @Override void begin(final long ptr) { }
        @Override void end(final long ptr) { }
    }
}

File: shankejdk-simple-sources/jdk/src/macosx/native/jobjc/src/core/native/NativeObjectLifecycleManager.m
/*
 * Copyright (c) 2011, 2012, Oracle and/or its affiliates. All rights reserved.
 */

#include "com_apple_jobjc_NativeObjectLifecycleManager.h"

#include <JavaNativeFoundation/JavaNativeFoundation.h>


/*
 * Class:     com_apple_jobjc_NativeObjectLifecycleManager
 * Method:    retainNativeObject
 * Signature: (J)V
 */
JNIEXPORT void JNICALL Java_com_apple_jobjc_NativeObjectLifecycleManager_retainNativeObject
(JNIEnv *env, jclass clazz, jlong ptr)
{
    if (ptr == 0L) return;
    CFRetain(jlong_to_ptr(ptr));
}

/*
 * Class:     com_apple_jobjc_NativeObjectLifecycleManager
 * Method:    releaseNativeObject
 * Signature: (J)V
 */
JNIEXPORT void JNICALL Java_com_apple_jobjc_NativeObjectLifecycleManager_releaseNativeObject
(JNIEnv *env, jclass clazz, jlong ptr)
{
    if (ptr == 0L) return;
    CFRelease(jlong_to_ptr(ptr));
}

/*
 * Class:     com_apple_jobjc_NativeObjectLifecycleManager
 * Method:    freeNativeObject
 * Signature: (J)V
 */
JNIEXPORT void JNICALL Java_com_apple_jobjc_NativeObjectLifecycleManager_freeNativeObject
(JNIEnv *env, jclass clazz, jlong ptr)
{
    if (ptr == 0L) return;
    free(jlong_to_ptr(ptr));
}
